icicle_v3/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,667 @@
+    } else {
+      segment.triangle_sum = P::zero();
+      segment.line_sum = P::zero();
+    }
+    segment.m_idx_in_segment = m_segment_size - 2;
+    segment.m_segment_mem_start = m_num_bkts * (i + 1) - m_segment_size + 1;
+  }
+}
+
+template <typename A, typename P>
+void Msm<A, P>::final_accumulator(std::shared_ptr<std::vector<BmSumSegment>>& segments_ptr, int idx_in_batch, P* result)
+{
+  // If it isn't the last MSM in the batch - run phase 3 on a separate thread to start utilizing the tasks manager on
+  // the next phase 1.
+  if (idx_in_batch == m_batch_size - 1) {
+    phase3_thread(segments_ptr, result);
+  } else {
+    m_p3_threads[idx_in_batch] = std::thread(&Msm<A, P>::phase3_thread, this, segments_ptr, result);
+  }
+}
+
+template <typename A, typename P>
+void Msm<A, P>::phase3_thread(std::shared_ptr<std::vector<BmSumSegment>> segments_ptr, P* result)
+{
+  auto& segments = *segments_ptr; // For readability
+  for (int i = 0; i < m_num_bms; i++) {
+    // Weighted sum of all the lines for each bm - summed in a similar fashion of the triangle sum of phase 2
+    P partial_sum = segments[m_num_bm_segments * (i + 1) - 1].line_sum;
+    P total_sum = segments[m_num_bm_segments * (i + 1) - 1].line_sum;
+    for (int j = m_num_bm_segments - 2; j > 0; j--) {
+      partial_sum = partial_sum + segments[m_num_bm_segments * i + j].line_sum;
+      total_sum = total_sum + partial_sum;
+    }
+    segments[m_num_bm_segments * i].line_sum = total_sum;
+
+    // Convert weighted lines sum to rectangles sum by doubling
+    int num_doubles = m_c - 1 - m_log_num_segments;
+    for (int k = 0; k < num_doubles; k++) {
+      segments[m_num_bm_segments * i].line_sum = P::dbl(segments[m_num_bm_segments * i].line_sum);
+    }
+
+    // Sum triangles within bm linearly
+    for (int j = 1; j < m_num_bm_segments; j++) {
+      segments[m_num_bm_segments * i].triangle_sum =
+        segments[m_num_bm_segments * i].triangle_sum + segments[m_num_bm_segments * i + j].triangle_sum;
+    }
+
+    // After which add the lines and triangle sums to one sum of the entire BM
+    if (m_num_bm_segments > 1) {
+      segments[m_num_bm_segments * i].triangle_sum =
+        segments[m_num_bm_segments * i].triangle_sum + segments[m_num_bm_segments * i].line_sum;
+    }
+  }
+
+  // Sum BM sums together
+  P final_sum = segments[(m_num_bms - 1) * m_num_bm_segments].triangle_sum;
+  for (int i = m_num_bms - 2; i >= 0; i--) {
+    // Multiply by the BM digit factor 2^c - i.e. c doublings
+    for (int j = 0; j < m_c; j++) {
+      final_sum = P::dbl(final_sum);
+    }
+    final_sum = final_sum + segments[m_num_bm_segments * i].triangle_sum;
+  }
+  *result = final_sum;
+}
+
+// None class functions below:
+
+/**
+ * @brief Function to check the MSM config is valid for calculating in CPU.
+ * @return - status if the config is supported or not.
+ */
+eIcicleError not_supported(const MSMConfig& conf)
+{
+  if (conf.is_async) { return eIcicleError::INVALID_DEVICE; }
```

Comment:
why? just be sync. It's not really an error.

Reply:
removed from the function